File: akka-docs/src/test/java/jdocs/cluster/SimpleClusterListener2.java
/*
 * Copyright (C) 2018-2019 Lightbend Inc. <https://www.lightbend.com>
 */

package jdocs.cluster;

import akka.actor.AbstractActor;
import akka.cluster.Cluster;
import akka.cluster.ClusterEvent.CurrentClusterState;
import akka.cluster.ClusterEvent.MemberEvent;
import akka.cluster.ClusterEvent.MemberUp;
import akka.cluster.ClusterEvent.MemberRemoved;
import akka.cluster.ClusterEvent.UnreachableMember;
import akka.event.Logging;
import akka.event.LoggingAdapter;

public class SimpleClusterListener2 extends AbstractActor {
  LoggingAdapter log = Logging.getLogger(getContext().getSystem(), this);
  //#join
  Cluster cluster = Cluster.get(getContext().getSystem());
  //#join

  //subscribe to cluster changes
  @Override
  public void preStart() {
    //#join
    cluster.join(cluster.selfAddress());
    //#join

    //#subscribe
    cluster.subscribe(getSelf(), MemberEvent.class, UnreachableMember.class);
    //#subscribe

    //#register-on-memberup
    cluster.registerOnMemberUp(
      () -> cluster.subscribe(getSelf(), MemberEvent.class, UnreachableMember.class)
    );
    //#register-on-memberup
  }

  //re-subscribe when restart
  @Override
  public void postStop() {
    cluster.unsubscribe(getSelf());
  }

  @Override
  public Receive createReceive() {
    return receiveBuilder()
      .match(CurrentClusterState.class, state -> {
        log.info("Current members: {}", state.members());
      })
      .match(MemberUp.class, mUp -> {
        log.info("Member is Up: {}", mUp.member());
      })
      .match(UnreachableMember.class, mUnreachable -> {
        log.info("Member detected as unreachable: {}", mUnreachable.member());
      })
      .match(MemberRemoved.class, mRemoved -> {
        log.info("Member is Removed: {}", mRemoved.member());
      })
      .match(MemberEvent.class, event -> {
        // ignore
      })
      .build();
  }
}

File: akka-docs/src/test/scala/docs/cluster/SimpleClusterListener2.scala
/*
 * Copyright (C) 2018-2019 Lightbend Inc. <https://www.lightbend.com>
 */

package docs.cluster

import akka.actor.{ Actor, ActorLogging }
import akka.cluster.Cluster
import akka.cluster.ClusterEvent._

class SimpleClusterListener2 extends Actor with ActorLogging {

  //#join
  val cluster = Cluster(context.system)
  //#join

  // subscribe to cluster changes, re-subscribe when restart
  override def preStart(): Unit = {
    //#join
    cluster.join(cluster.selfAddress)
    //#join

    //#subscribe
    cluster.subscribe(self, classOf[MemberEvent], classOf[UnreachableMember])
    //#subscribe

    //#register-on-memberup
    cluster.registerOnMemberUp {
      cluster.subscribe(self, classOf[MemberEvent], classOf[UnreachableMember])
    }
    //#register-on-memberup
  }
  override def postStop(): Unit = cluster.unsubscribe(self)

  def receive = {
    case state: CurrentClusterState ⇒
      log.info("Current members: {}", state.members.mkString(", "))
    case MemberUp(member) ⇒
      log.info("Member is Up: {}", member.address)
    case UnreachableMember(member) ⇒
      log.info("Member detected as unreachable: {}", member)
    case MemberRemoved(member, previousStatus) ⇒
      log.info(
        "Member is Removed: {} after {}",
        member.address, previousStatus)
    case _: MemberEvent ⇒ // ignore
  }
}
